File: LLD/Design-Patterns/AbstractFactory/code/AbstractFactory.java
// Abstract Factory Pattern — cross-platform UI widget families.
// Compile: javac AbstractFactory.java
// Run:     java AbstractFactory

interface Button {
    void render();
}

interface Checkbox {
    void render();
}

class WindowsButton implements Button {
    public void render() { System.out.println("Rendering a Windows-style button."); }
}

class WindowsCheckbox implements Checkbox {
    public void render() { System.out.println("Rendering a Windows-style checkbox."); }
}

class MacButton implements Button {
    public void render() { System.out.println("Rendering a Mac-style button."); }
}

class MacCheckbox implements Checkbox {
    public void render() { System.out.println("Rendering a Mac-style checkbox."); }
}

interface UIFactory {
    Button createButton();
    Checkbox createCheckbox();
}

class WindowsFactory implements UIFactory {
    public Button createButton() { return new WindowsButton(); }
    public Checkbox createCheckbox() { return new WindowsCheckbox(); }
}

class MacFactory implements UIFactory {
    public Button createButton() { return new MacButton(); }
    public Checkbox createCheckbox() { return new MacCheckbox(); }
}

public class AbstractFactory {
    static void renderUI(UIFactory factory) {
        Button button = factory.createButton();
        Checkbox checkbox = factory.createCheckbox();
        button.render();
        checkbox.render();
    }

    public static void main(String[] args) {
        System.out.println("-- Windows family --");
        renderUI(new WindowsFactory());

        System.out.println("-- Mac family --");
        renderUI(new MacFactory());
    }
}

File: LLD/Design-Patterns/AbstractFactory/code/abstract_factory.py
"""
Abstract Factory Pattern — cross-platform UI widget families.
Run: python abstract_factory.py
"""

from abc import ABC, abstractmethod


class Button(ABC):
    @abstractmethod
    def render(self):
        ...


class Checkbox(ABC):
    @abstractmethod
    def render(self):
        ...


class WindowsButton(Button):
    def render(self):
        print("Rendering a Windows-style button.")


class WindowsCheckbox(Checkbox):
    def render(self):
        print("Rendering a Windows-style checkbox.")


class MacButton(Button):
    def render(self):
        print("Rendering a Mac-style button.")


class MacCheckbox(Checkbox):
    def render(self):
        print("Rendering a Mac-style checkbox.")


class UIFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        ...

    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        ...


class WindowsFactory(UIFactory):
    def create_button(self) -> Button:
        return WindowsButton()

    def create_checkbox(self) -> Checkbox:
        return WindowsCheckbox()


class MacFactory(UIFactory):
    def create_button(self) -> Button:
        return MacButton()

    def create_checkbox(self) -> Checkbox:
        return MacCheckbox()


def render_ui(factory: UIFactory):
    button = factory.create_button()
    checkbox = factory.create_checkbox()
    button.render()
    checkbox.render()


if __name__ == "__main__":
    print("-- Windows family --")
    render_ui(WindowsFactory())

    print("-- Mac family --")
    render_ui(MacFactory())
